### PR TITLE
mypy fix index_map.py

### PR DESCRIPTION
**Pull request by stevebachmeier**
### Description

- *Category*:  other
- *JIRA issue*: https://jira.ihme.washington.edu/browse/MIC-5346

### CODE REVIEW

File: src/vivarium/framework/randomness/index_map.py [hunk 1]
```diff
@@ -62,7 +64,7 @@ def update(self, new_keys: pd.DataFrame, clock_time: pd.Timestamp) -> None:
         final_mapping = final_mapping.sort_index(level=self.SIM_INDEX_COLUMN)
         self._map = final_mapping
 
-    def _parse_new_keys(self, new_keys: pd.DataFrame) -> Tuple[pd.MultiIndex, pd.MultiIndex]:
+    def _parse_new_keys(self, new_keys: pd.DataFrame) -> tuple[pd.Index[Any], pd.Index[Any]]:
```
[the remark this answers is not in the document]
Comment (stevebachmeier) [reply src/vivarium/framework/randomness/index_map.py:67]: As discussed, my thought was that these can be either MultiIndex or Index, and then if it's a MultiIndex, they really can be anything.

File: src/vivarium/framework/randomness/index_map.py [hunk 2]
```diff
@@ -88,12 +90,12 @@ def _parse_new_keys(self, new_keys: pd.DataFrame) -> Tuple[pd.MultiIndex, pd.Mul
         if self._map is None:
             final_mapping_index = new_mapping_index
         else:
-            final_mapping_index = self._map.index.append(new_mapping_index)
+            final_mapping_index = self._map.index.append(new_mapping_index)  # type: ignore [no-untyped-call]
```
[the remark this answers is not in the document]
Comment (stevebachmeier) [reply src/vivarium/framework/randomness/index_map.py:93]: As discussed:
1. We handle self._map == None on line 90
2. we looked a bit into pd.Index.append. It just looks like it's only partially typed or something?

File: src/vivarium/framework/randomness/index_map.py [hunk 3]
```diff
@@ -183,11 +187,13 @@ def _hash(self, keys: pd.Index, salt: int = 0) -> pd.Series:
...
                 # introduces is pretty trivial.
                 out *= np.power(p, self._digit(column, idx))
-            new_map += out + salt
+            new_map += out + salt_series
 
         return new_map % len(self)
 
-    def _convert_to_ten_digit_int(self, column: pd.Series) -> pd.Series:
+    def _convert_to_ten_digit_int(
+        self, column: pd.Series[Union[datetime.datetime, int, float]]
```
[the remark this answers is not in the document]
Comment (stevebachmeier) [reply src/vivarium/framework/randomness/index_map.py:195]: Yeah, trying ClockTime gave `error: Type argument "Timestamp | datetime | Number | int | float" of "Series" must be a subtype of "str | bytes | date | time | bool | <10 more items>"  [type-var]`. Unsure if that's what we'd expect...

File: tests/framework/randomness/test_index_map.py [hunk 4]
```diff
@@ -61,13 +61,6 @@ def map_size_and_hashed_values(request):
     return len(m), m._hash(keys)
 
 
-def test_digit_scalar():
```
[the remark this answers is not in the document]
Comment (stevebachmeier) [reply tests/framework/randomness/test_index_map.py:64]: As discussed, we really don't support scalars, only pd.Series'

File: src/vivarium/framework/randomness/index_map.py [hunk 5]
```diff
@@ -12,11 +12,12 @@
...
 from __future__ import annotations
 
-import datetime
-from typing import Any, Optional, Union
+from datetime import datetime
+from typing import Any
 
 import numpy as np
 import pandas as pd
+from pandas.api import types as pdtypes
```
Comment (rmudambi) [inline src/vivarium/framework/randomness/index_map.py:20]: I've seen that `numpy.typing` is often imported as `import numpy.typing as npt`. Is there a similar convention for this? If not, do we prefer `pdtypes` or `pd_types`?

**Review by rmudambi — APPROVED** [review]
Let's wait to merge/release until my forthcoming PR get approved.

File: src/vivarium/framework/randomness/index_map.py (hunk 5 above)
Comment (stevebachmeier) [reply src/vivarium/framework/randomness/index_map.py:20]: Oh i prefer pdt then

**Review by albrja — APPROVED** [review]

**Review by patricktnast — APPROVED** [review]

**Pull request merged**